src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -536,15 +537,32 @@ internal unsafe ValueTask ConnectAsync(CancellationToken cancellationToken = def
                 // We don't have way how to set separate SNI and name for connection at this moment.
                 // If the name is actually IP address we can use it to make at least some cases work for people
                 // who want to bypass DNS but connect to specific virtual host.
-                if (!string.IsNullOrEmpty(_state.TargetHost) && !dnsHost.Equals(_state.TargetHost, StringComparison.InvariantCultureIgnoreCase) && IPAddress.TryParse(dnsHost, out IPAddress? address))
+                if (!dnsHost.Equals(_state.TargetHost, StringComparison.InvariantCultureIgnoreCase) && !string.IsNullOrEmpty(_state.TargetHost))
                 {
-                    // This is form of IPAddress and _state.TargetHost is set to different string
-                    Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
-                    MsQuicParameterHelpers.SetIPEndPointParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_CONN_REMOTE_ADDRESS, new IPEndPoint(address, port));
                     targetHost = _state.TargetHost!;
+                    if (IPAddress.TryParse(dnsHost, out IPAddress? address))
+                    {
+                        // This is form of IPAddress and _state.TargetHost is set to different string
+                        Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
+                        MsQuicParameterHelpers.SetIPEndPointParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_CONN_REMOTE_ADDRESS, new IPEndPoint(address, port));
+                    }
+                    else
+                    {
+                        IPAddress[] addresses = Dns.GetHostAddressesAsync(dnsHost, cancellationToken).GetAwaiter().GetResult();
+                        cancellationToken.ThrowIfCancellationRequested();
+                        if (addresses.Length == 0)
+                        {
+                            throw new SocketException((int)SocketError.HostNotFound);
```

Comment:
`QuicException`?

Reply:
> fail to resolve may not be IOException

Why?

We should then reconsider inheriting QuicException from IOException, but we shouldn't throw SocketException from S.N.Quic.

Reply:
as I mentioned we may do it anyway if Dns.GetHostAddressesAsync fails. Unless you want to catch it and throw something else in both cases. We can certainly do it as well.

Reply:
Baaah, I'm not sure, let's keep the `SocketException` for now to be consistent with `NameResolution` class. We can revisit this in the future if it'll cause confusion.